mountpoint-s3/tests/fuse_tests/readdir_test.rs

```diff
@@ -30,38 +33,112 @@ fn prepare_fs(mut put_object_fn: PutObjectFn, map: &HashMap<String, File>) {
     }
 }
 
-fn readdir_1<F>(creator_fn: F, prefix: &str)
+fn readdir<F>(creator_fn: F, prefix: &str)
 where
     F: FnOnce(&str, S3FilesystemConfig) -> (TempDir, BackgroundSession, PutObjectFn),
 {
-    let readdir_size = 5usize;
+    let readdir_size = 5;
     let filesystem_config = S3FilesystemConfig {
         readdir_size,
         ..Default::default()
     };
 
     let mut map = HashMap::new();
+    let mut expected_list = Vec::new();
     for i in 0..readdir_size * 4 {
-        map.insert(format!("file{i}"), File::new((i % 256) as u8, 10 * i));
+        let random_str = Alphanumeric.sample_string(&mut rand::thread_rng(), 5);
+        let file_name = format!("file_{random_str}_{i}");
+        map.insert(file_name.clone(), File::new((i % 256) as u8, 10 * i));
+        expected_list.push(file_name);
     }
+    expected_list.sort();
+
     let (mount_point, _session, put_object_fn) = creator_fn(prefix, filesystem_config);
 
     prepare_fs(put_object_fn, &map);
 
     let dir = fs::read_dir(mount_point.path()).unwrap();
-    let mut files: Vec<_> = dir.map(|f| f.unwrap()).collect();
+    let dirs: Vec<_> = dir.map(|f| f.unwrap()).collect();
+    assert_eq!(
+        dirs.iter()
+            .map(|f| f.path().file_name().unwrap().to_str().unwrap().to_owned())
+            .collect::<Vec<_>>(),
+        expected_list
+    );
+}
 
-    for f in files.drain(..) {
-        let path = f.path();
-        let name = path.file_name().unwrap().to_str().unwrap();
-        assert!(map.remove(name).is_some());
+fn readdir_while_writing<F>(creator_fn: F, prefix: &str)
+where
+    F: FnOnce(&str, S3FilesystemConfig) -> (TempDir, BackgroundSession, PutObjectFn),
+{
+    let readdir_size = 5;
+    let filesystem_config = S3FilesystemConfig {
+        readdir_size,
+        ..Default::default()
+    };
+
+    let mut map = HashMap::new();
+    let mut expected_list = Vec::new();
+    for i in 0..readdir_size * 4 {
+        let random_str = Alphanumeric.sample_string(&mut rand::thread_rng(), 5);
```

Comment:
We should use a fixed seed for the RNG here (and everywhere we use `rand::thread_rng()`) so test failures are reproducible. Then we should probably run this whole test in a loop, at least in the mock case, so we can get coverage of different random seeds.

Reply:
Updated these tests.